### PR TITLE
[FIX] Switch from esprima to espree

### PR DESCRIPTION
**Pull request by codeworrior**
This is work in progress and can't be submitted as the used espree
version (8.0.0-beta) is not a release version and it requires a newer
NodeJS min version than the tooling.

espree 8.0.0-beta was chosen as it switches the default for the
ecmaVersion option from 5 (7.3.1) to latest.

Before submitting, it also should be decided whether all parsing code
should be routed through some helper utility, to simplify future updates
/ changes of the parser. At the same time, we should decide about the
use of the sourceType and the ecmaVersion options. Should we expose them
in ui5.yaml? Or should we rather leave that to the linting step?

Fixes https://github.com/SAP/ui5-tooling/issues/354 .

**Thank you for your contribution!** 🙌

To get it merged faster, kindly review the checklist below:

## Pull Request Checklist
- [ ] Reviewed the [Contributing Guidelines](https://github.com/SAP/ui5-tooling/blob/master/CONTRIBUTING.md#-contributing-code)
    + Especially the [How to Contribute](https://github.com/SAP/ui5-tooling/blob/master/CONTRIBUTING.md#how-to-contribute) section 
- [ ] [No merge commits](https://github.com/SAP/ui5-tooling/blob/master/docs/Guidelines.md#no-merge-commits)
- [ ] [Correct commit message style](https://github.com/SAP/ui5-tooling/blob/master/docs/Guidelines.md#commit-message-style)

### CODE REVIEW

**Comment by RandomByte** [comment]
> At the same time, we should decide about the use of the sourceType and the ecmaVersion options. Should we expose them in ui5.yaml? Or should we rather leave that to the linting step?

As discussed earlier today, we should always parse the latest ECMAScript version (just like browsers do). Projects should use linters to enforce specific versions.  
We should always assume sourceType "script" for now. In the future we might detect modules based on file suffix or package.json configuration.

To make this part of a non-major UI5 Tooling release we would need to use espree v6 as all later versions require Node.js >10.

**Comment by codeworrior** [comment]
Going with espree ^6.0.0 however, we would miss at least the following syntax features:
* [75e80bc](https://github.com/eslint/espree/commit/75e80bc1b5a100f64b09b2c7bf47c5168417d890) Update: support ?? operator, import.meta, and export * as ns
* [872645c](https://github.com/eslint/espree/commit/872645cea0bee08960b93c097f84153d44b44d7f) Update: support optional chaining
* [bd0a405](https://github.com/eslint/espree/commit/bd0a405ffbe4962bcf5aa225ed3861a6a15cb827) Update: support logical assignment and numeric separators

where the second one hurts most.

**Comment by RandomByte** [comment]
Still better than nothing, right? So maybe go with v6 for now and upgrade to v8 with UI5 Tooling 3.0

**Comment by RandomByte** [comment]
Validated 7ea1a69 successfully in a smoke test run and locally by building the webc libraries 👍

**Comment by codeworrior** [comment]
Are you okay with the `parseJS` module? In the tests, I found it a bit strange that the module has the same name as one of its named exports. If someone has a better proposal, I would be glad to rename it. 

* Module: `parseUtil`
* Exports: 'parseJS`, `Syntax`

**Comment by RandomByte** [comment]
Yeah, parseUtil seems to be a better name for the module

**Comment by RandomByte** [comment]
I'll do the rename

**Comment by RandomByte** [comment]
Decided on parseUtil**s** to align it with the already existing ASTUtils.

What about `parseJs` instead of `parseJS` 🤔 🙈

**Comment by codeworrior** [comment]
Reg. `parseJs` you do what you have to, but for me it looks wrong...

**Comment by RandomByte** [comment]
Well there is already `JSTokenizer.parseJS`, so let's rather stick with it here too.

Anything left here?

**Review by RandomByte — APPROVED** [review]

File: test/lib/lbt/analyzer/SmartTemplateAnalyzer.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ const test = require("ava");
 const SmartTemplateAnalyzer = require("../../../../lib/lbt/analyzer/SmartTemplateAnalyzer");
 const ModuleInfo = require("../../../../lib/lbt/resources/ModuleInfo");
 const sinon = require("sinon");
-const esprima = require("esprima");
+const parseJS = require("../../../../lib/lbt/utils/parseUtils");
```
Comment (matz3) [inline test/lib/lbt/analyzer/SmartTemplateAnalyzer.js:5]: `parseUtils`

File: test/lib/lbt/analyzer/FioriElementsAnalyzer.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 const test = require("ava");
 const FioriElementsAnalyzer = require("../../../../lib/lbt/analyzer/FioriElementsAnalyzer");
 const sinon = require("sinon");
-const esprima = require("esprima");
+const parseJS = require("../../../../lib/lbt/utils/parseUtils");
```
Comment (matz3) [inline test/lib/lbt/analyzer/FioriElementsAnalyzer.js:4]: `parseUtils`

File: test/lib/lbt/analyzer/SmartTemplateAnalyzer.js (hunk 1 above)
Comment (RandomByte) [reply test/lib/lbt/analyzer/SmartTemplateAnalyzer.js:5]: Thanks, done in https://github.com/SAP/ui5-builder/pull/619

File: test/lib/lbt/analyzer/FioriElementsAnalyzer.js (hunk 2 above)
Comment (RandomByte) [reply test/lib/lbt/analyzer/FioriElementsAnalyzer.js:4]: Done in https://github.com/SAP/ui5-builder/pull/619

**Pull request merged**